### ISSUE TITLE
Optional programs are no longer optional

### ISSUE THREAD

**Issue by QuLogic** [jbarlow83/OCRmyPDF#360]
When `unpaper` is not installed, even the tests that do not use it fail to run; the program crashes instead of of skipping it:
```
___________________________ test_no_glyphless_weave ____________________________

resources = PosixPath('/builddir/build/BUILD/ocrmypdf-8.2.0/tests/resources')
outdir = PosixPath('/tmp/pytest-of-mockbuild/pytest-0/test_no_glyphless_weave0')

    def test_no_glyphless_weave(resources, outdir):
        pdf = pikepdf.open(resources / 'francais.pdf')
        pdf_aspect = pikepdf.open(resources / 'aspect.pdf')
        pdf_cmyk = pikepdf.open(resources / 'cmyk.pdf')
        pdf.pages.extend(pdf_aspect.pages)
        pdf.pages.extend(pdf_cmyk.pages)
        pdf.save(outdir / 'test.pdf')
    
        env = os.environ.copy()
        env['_OCRMYPDF_MAX_OPEN_PAGE_PDFS'] = '2'
        check_ocrmypdf(
            outdir / 'test.pdf',
            outdir / 'out.pdf',
            '--deskew',
            '--tesseract-timeout',
            '0',
>           env=env,
        )

tests/test_weave.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input_file = PosixPath('/tmp/pytest-of-mockbuild/pytest-0/test_no_glyphless_weave0/test.pdf')
output_file = PosixPath('/tmp/pytest-of-mockbuild/pytest-0/test_no_glyphless_weave0/out.pdf')
env = {'BASH_ENV': '/usr/share/lmod/lmod/init/bash', 'BASH_FUNC_ml%%': '() {  eval $($LMOD_DIR/ml_cmd "$@")\n}', 'BASH_FUNC_module%%': '() {  eval $($LMOD_CMD bash "$@") && eval $(${LMOD_SETTARG_CMD:-:} -s sh)\n}', 'CONFIG_SITE': 'NONE', ...}
args = ('--deskew', '--tesseract-timeout', '0')
p = <subprocess.Popen object at 0x7f73b87bd6a0>, out = ''

    @pytest.helpers.register
    def check_ocrmypdf(input_file, output_file, *args, env=None):
        """Run ocrmypdf and confirmed that a valid file was created"""
    
        p, out, err = run_ocrmypdf(input_file, output_file, *args, env=env)
        # ensure py.test collects the output, use -s to view
        print(err, file=sys.stderr)
>       assert p.returncode == 0
E       assert 1 == 0
E        +  where 1 = <subprocess.Popen object at 0x7f73b87bd6a0>.returncode

tests/conftest.py:155: AssertionError
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/exec/__init__.py", line 38, in get_version
    check=True,
  File "/usr/lib64/python3.7/subprocess.py", line 472, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib64/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/usr/lib64/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'unpaper': 'unpaper'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/exec/__init__.py", line 150, in check_external_program
    found_version = version_checker()
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/exec/unpaper.py", line 41, in version
    return get_version('unpaper')
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/exec/__init__.py", line 44, in get_version
    ) from e
ocrmypdf.exceptions.MissingDependencyError: Could not find program 'unpaper' on the PATH

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/__main__.py", line 1183, in <module>
    sys.exit(run_pipeline())
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/__main__.py", line 1093, in run_pipeline
    check_dependency_versions(_log)
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/__main__.py", line 1057, in check_dependency_versions
    optional=True,
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/exec/__init__.py", line 152, in check_external_program
    error_missing_program(log, program, package, optional)
  File "/builddir/build/BUILDROOT/ocrmypdf-8.2.0-1.fc30.x86_64/usr/lib/python3.7/site-packages/ocrmypdf/exec/__init__.py", line 132, in error_missing_program
    log.error(missing_program.format(**locals()), file=sys.stderr)
  File "/usr/lib/python3.7/site-packages/ruffus/proxy_logger.py", line 292, in error
    return self._callmethod('error', args, kwargs)
  File "/usr/lib64/python3.7/multiprocessing/managers.py", line 795, in _callmethod
    conn.send((self._id, methodname, args, kwds))
  File "/usr/lib64/python3.7/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib64/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot serialize '_io.TextIOWrapper' object
```

**To Reproduce**
Run tests without `unpaper` installed.

**Expected behavior**
Running with `unpaper` should not fail unless, e.g., `--clean` or `--clean-final`, are supplied.

**System:**
- OS: Fedora Rawhide
- OCRmyPDF Version: 8.2.0

**Comment by QuLogic** [comment]
Though maybe this is a bug in `ruffus` or `multiprocessing`...

**Comment by QuLogic** [comment]
Passing `file` to standard Python `logging` does not work, so I'm not sure it should work for `ruffus.proxy_logger` either.

**Comment by jbarlow83** [comment]
It was just a regression on my part, the other libraries are not affected.
Fixed in v8.2.1

**Comment by jbarlow83** [comment]
Correction: fixed in v8.2.2

**Comment by QuLogic** [comment]
Great! It [works well](https://bodhi.fedoraproject.org/updates/FEDORA-2019-f039157e9e) now, and I don't even need to patch it to skip unpaper tests anymore.
